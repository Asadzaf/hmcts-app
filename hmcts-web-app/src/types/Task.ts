type Task = {
    id: number | null;
    title: string;
    description: string | null;
    status: string;
    dueDate: string;
}

export default Task;