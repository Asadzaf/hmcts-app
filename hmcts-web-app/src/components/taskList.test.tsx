import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import TaskList from "./taskList";
import Task from "../types/Task";


const renderTaskList = (tasks: Task[]) => {
    render(
        <MemoryRouter>
            <TaskList data={tasks} />
        </MemoryRouter>
    );
};


describe("TaskList component", () => {
    const mockTasks: Task[] = [
        {
            id: 1001,
            title: "Test Task 1",
            description: "Description 1",
            status: "ACTIVE",
            dueDate: "2024-06-01",
        },
        {
            id: 6049,
            title: "Test Task 2",
            description: "Description 2",
            status: "INACTIVE",
            dueDate: "2024-07-01",
        },
    ];

    it("renders all tasks", () => {
        renderTaskList(mockTasks);

        // checks length = 2 as task title is shown twice, once in card title and once in card content
        expect(screen.getAllByText("Test Task 1").length).toEqual(2);
        expect(screen.getAllByText("Test Task 2").length).toEqual(2);
    });

    it("renders task details correctly", () => {
        renderTaskList(mockTasks);
        expect(screen.getAllByText("ID").length).toBe(2);
        expect(screen.getByText("1001")).toBeInTheDocument();
        expect(screen.getByText("6049")).toBeInTheDocument();
        expect(screen.getByText("Description 1")).toBeInTheDocument();
        expect(screen.getByText("Description 2")).toBeInTheDocument();
        expect(screen.getByText("ACTIVE")).toBeInTheDocument();
        expect(screen.getByText("INACTIVE")).toBeInTheDocument();
        expect(screen.getByText("2024-06-01")).toBeInTheDocument();
        expect(screen.getByText("2024-07-01")).toBeInTheDocument();
    });

    it("renders correct Update/Delete links", () => {
        renderTaskList(mockTasks);
        const updateLinks = screen.getAllByRole("link", { name: "Update" });
        updateLinks.forEach(link => {
            expect(link).toHaveAttribute("href", "/viewTask");
        });
    });

    it("renders nothing if data is empty", () => {
        renderTaskList([]);
        expect(screen.queryByText("ID")).not.toBeInTheDocument();
    });

    test('renders taskList correctly via snapshot', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <TaskList data={mockTasks} />
            </MemoryRouter>
        )

        expect(asFragment()).toMatchSnapshot();
    });
});
