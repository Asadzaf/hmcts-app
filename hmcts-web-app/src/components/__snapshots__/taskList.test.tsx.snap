// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList component renders taskList correctly via snapshot 1`] = `
<DocumentFragment>
  <div
    class="govuk-summary-card"
  >
    <div
      class="govuk-summary-card__title-wrapper"
    >
      <h2
        class="govuk-summary-card__title"
      >
        Test Task 1
      </h2>
      <div
        class="govuk-summary-card__actions"
      >
        <li
          class="govuk-summary-card__action"
        >
          <a
            class="govuk-link"
            data-discover="true"
            href="/viewTask"
          >
            Update
          </a>
        </li>
        <li
          class="govuk-summary-card__action"
        >
          <a
            class="govuk-link"
            data-discover="true"
            href="/deleteTask"
          >
            Delete
          </a>
        </li>
      </div>
    </div>
    <div
      class="govuk-summary-card__content"
    >
      <div
        class="govuk-summary-list"
      >
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            ID
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            1001
          </dd>
        </div>
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            Task Title
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            Test Task 1
          </dd>
        </div>
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            Description
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            Description 1
          </dd>
        </div>
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            Status
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            ACTIVE
          </dd>
        </div>
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            Due Date
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            2024-06-01
          </dd>
        </div>
      </div>
    </div>
  </div>
  <div
    class="govuk-summary-card"
  >
    <div
      class="govuk-summary-card__title-wrapper"
    >
      <h2
        class="govuk-summary-card__title"
      >
        Test Task 2
      </h2>
      <div
        class="govuk-summary-card__actions"
      >
        <li
          class="govuk-summary-card__action"
        >
          <a
            class="govuk-link"
            data-discover="true"
            href="/viewTask"
          >
            Update
          </a>
        </li>
        <li
          class="govuk-summary-card__action"
        >
          <a
            class="govuk-link"
            data-discover="true"
            href="/deleteTask"
          >
            Delete
          </a>
        </li>
      </div>
    </div>
    <div
      class="govuk-summary-card__content"
    >
      <div
        class="govuk-summary-list"
      >
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            ID
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            6049
          </dd>
        </div>
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            Task Title
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            Test Task 2
          </dd>
        </div>
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            Description
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            Description 2
          </dd>
        </div>
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            Status
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            INACTIVE
          </dd>
        </div>
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            Due Date
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            2024-07-01
          </dd>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
